### PR TITLE
feat(gateway): implement budget principal taxonomy

### LINKED ISSUE #106: [Feature]: Rework budget hierarchy and owner taxonomy
## Summary

Revisit the budget owner hierarchy and taxonomy so spend controls can distinguish personal user limits, softer team-level controls, team-owned/service-account usage, and model-specific limits per user. This matters because the current budget engine treats user-owned and team-owned API keys as separate owner scopes, but team-owned keys cannot constrain individual team members and service-account semantics are not first-class.

## Problem

Current budget enforcement is organized around two API key owner kinds only: `user` and `team`. User-owned API keys use `user_budgets`; team-owned API keys use `team_budgets`; team ownership scope is recorded as `team:<team_id>:actor:none`.

That leaves several policy gaps:

- personal budgets are the only reliable way to prevent one user from consuming disproportionate spend;
- team budgets are useful for aggregate governance, but are too coarse to be the primary fairness control;
- team-owned/API-service credentials are currently just team-owned keys, not a distinct service-account owner or budget scope;
- model costs vary enough that a single user budget cannot express “cheap models broadly, expensive models carefully”; and
- the budget guard has duplicated user/team logic that will get harder to harden if more budget dimensions are added directly.

## Current Risk

- A single team member can use a team-owned key and consume most or all of a team budget without a personal cap.
- Service-account-style traffic is indistinguishable from general team-owned traffic unless operators create synthetic teams or rely on the reserved `system-legacy` team behavior.
- Expensive models can drain the same budget pool as low-cost models, even when the expensive model is not appropriate for most tasks.
- Adding new budget dimensions ad hoc could make enforcement ordering, reporting, alerts, and tests inconsistent.
- Related work in #25, #92, #100, and #101 may touch adjacent policy/reporting behavior without a clear budget taxonomy to align to.

## Desired Guarantee

Budget enforcement and reporting have a clear hierarchy and owner taxonomy where personal user budgets are the primary fairness control, team budgets provide aggregate guardrails, service-account/team-owned traffic has an explicit policy decision, and optional per-user model budgets can limit high-cost model usage without inventing a second accounting source.

## Scope (This Issue)

1. Define the budget hierarchy and enforcement order for:
   - personal user budgets;
   - team-level aggregate budgets;
   - team-owned or service-account budgets;
   - per-user model budgets.
2. Decide whether “service account” should become a first-class gateway API key owner kind or remain modeled as a team-owned key with clearer conventions.
3. Refactor the budget guard around reusable budget-scope evaluation instead of duplicating user/team branches as more dimensions are added.
4. Extend persistence/API/reporting contracts only as needed for the selected taxonomy.
5. Add tests that prove enforcement order, duplicate-request idempotency, unpriced/usage-missing behavior, and reporting alignment across budget dimensions.
6. Document the budget hierarchy, owner taxonomy, and migration/rollout guidance.

Out of scope:

- Implementing generic request rate limits or quotas from #25.
- Expanding provider-specific cache/reasoning/modality accounting from #92 except where model-budget tests need existing priced ledger rows.
- Provider breakdown reporting from #100.
- Timezone-aware windows from #101, though the design should use the same eventual window service.

## Acceptance Criteria

- [ ] A documented budget hierarchy explains which budget scopes are checked for user-owned, team-owned, and service-account-style traffic.
- [ ] Personal user budgets are enforced before or at least independently from aggregate team budgets for traffic attributable to a user.
- [ ] Team budgets are treated as aggregate guardrails rather than the only fairness control for team member traffic.
- [ ] Service-account/team-owned traffic has an explicit, tested policy: either first-class service-account owner budgets or documented team-owned service-account convention.
- [ ] Per-user model budgets can limit spend for specific gateway/upstream model identities without breaking the canonical `usage_cost_events` ledger.
- [ ] Budget enforcement remains idempotent on `(request_id, ownership_scope_key)` and does not charge duplicate requests twice.
- [ ] `unpriced` and `usage_missing` rows remain visible in reporting but do not count toward charged budget spend.
- [ ] Tests cover budget ordering, hard-limit rejection, soft/aggregate behavior where applicable, model-specific limits, and backend parity for any new persistence.
- [ ] Docs and ADRs are updated if the chosen taxonomy changes owner semantics or schema shape.

## Suggested Implementation Areas

- `crates/gateway-core/src/domain.rs`
- `crates/gateway-core/src/traits.rs`
- `crates/gateway-service/src/budget_guard.rs`
- `crates/gateway-service/src/service.rs`
- `crates/gateway-service/src/admin_api_keys.rs`
- `crates/gateway-store/src/libsql_store/budgets.rs`
- `crates/gateway-store/src/postgres_store/budgets.rs`
- `crates/gateway-store/migrations/`
- `crates/gateway/src/http/spend.rs`
- `crates/gateway/src/config.rs`
- `docs/operations/budgets-and-spending.md`
- `docs/reference/data-relationships.md`
- `docs/adr/`

## Notes

Issue scan found related but narrower work:

- #25 covers request quotas/rate limits and should stay separate from spend budgets.
- #92 covers richer usage accounting semantics that may affect model-budget accuracy.
- #100 covers provider spend reporting.
- #101 covers timezone-aware budget windows.
- #28 and #30 delivered the current live spend/budget control plane.

Current codebase findings:

- `ApiKeyOwnerKind` currently supports only `user` and `team`.
- Provider `auth.mode: service_account` is a provider credential mode, not a gateway caller identity.
- Config-seeded API keys are currently forced to the reserved `system-legacy` team.
- Runtime usage ledger rows already have `actor_user_id`, but request recording passes `None` today.
- Team ownership scope currently uses `team:<team_id>:actor:none`.
- Budgets currently exist only in `user_budgets` and `team_budgets`.

### PR DESCRIPTION
**Pull request by ahstn**
## Description

Implements the Issue 106 budget principal taxonomy pivot. Spend-bearing principals are now human users and service accounts only, with optional human user model budgets. Team budgets and team-owned runtime API keys are removed from runtime/API/UI paths, with legacy handling constrained to destructive V28 migration cleanup.

- Adds generic budget scope domain types, scope-key construction, repository methods, store implementations, and budget enforcement ordering.
- Adds V28 libsql/Postgres migrations for the generic `budgets` table, backfill, team-budget cleanup, and API-key owner constraints.
- Replaces path-specific budget admin routes with generic typed-scope budget routes.
- Enforces active service-account budgets before service-account API keys can be created or used.
- Reworks Spend Controls to User Budgets, Service Account Budgets, and User Model Budgets.
- Removes team owner filters/surfaces from spend reporting and API key admin payloads.
- Updates seeded API key config to require explicit service account, team association, and active budget.
- Regenerates OpenAPI/admin types.
- Splits documentation into operator internals and user-facing budget setup docs, plus ADR updates.

Why this approach was chosen:

The latest issue comment superseded the original team-budget design. A generic budget scope API keeps enforcement and admin mutation contracts cohesive while avoiding team fallbacks or free-form scope-key editing. Service-account budget activation is enforced at credential creation/auth time so non-human spend cannot bypass a declared budget.

Risks, tradeoffs, and alternatives considered:

- V28 intentionally deletes legacy team budget, team-owned key, and team-owned ledger data because the issue states there is no production accounting history to preserve.
- Hard-limit enforcement remains best-effort under concurrency; reservations are still out of scope.
- Postgres-specific smoke tasks were not run locally because no `TEST_POSTGRES_URL`/local Postgres test environment was available.

Fixes #106

## Release Readiness Checklist

- [x] `mise run lint`
- [x] `mise run test`
- [ ] If this PR touches runtime, store, migration, or release behavior: `mise run check-rust-postgres`
- [ ] If this PR touches runtime, store, migration, or release behavior: `mise run test-rust-postgres`
- [ ] If this PR touches runtime, store, migration, or release behavior: `mise run test-gateway-postgres-smoke`

Additional verification run:

- [x] `mise run admin-contract-generate`
- [x] `mise run admin-contract-check`
- [x] `mise -C docs run check`
- [x] `bun run --cwd crates/admin-ui/web build`

### CODE REVIEW

**Comment by ahstn** [comment]
## Multi-agent review summary

Ran `code-review-and-quality` as parallel axis passes against this PR and combined it with a separate `code-simplification` pass.

### Fixed in the latest amend (`d738e563`)

- Correctness: fixed the Postgres all-owner spend aggregate by restoring the missing `UNION ALL` between user and service-account rollups.
- Security/correctness: duplicate `request_id` values for the same spend owner now fail before provider execution and again during usage recording instead of allowing provider execution with skipped ledger insertion.
- Maintainability/simplification: removed the now-dead `BudgetGuardDisposition::Duplicate` path and made usage recording success-or-error.
- Maintainability: replaced the local ownership scope-key helper with the shared `budget_scopes::usage_ownership_scope_key` helper.
- Performance: added V28 ledger indexes for user-model budget checks: `(user_id, model_id, occurred_at)` and `(user_id, TRIM(upstream_model), occurred_at)`.

### Remaining non-blocking follow-ups

- `crates/gateway/src/http/spend.rs`: budget overview assembly still has N+1 query shape; a future store-level aggregate would scale better for large tenants.
- `crates/gateway/src/http/spend.rs`: budget admin orchestration still lives heavily in the HTTP layer; moving more of it to `gateway-service` would improve boundaries.
- Admin contract shape: `BudgetScopeRequest` / response scope objects are still nullable field bags rather than a generated tagged union/`oneOf` shape.
- Spend Controls UI remains a large page component; splitting repeated budget table/form sections would improve reviewability.
- Owner aggregate SQL still duplicates similar blocks across filtered/all-owner paths and libsql/Postgres backends.

### Verification

- `cargo check -p gateway-service -p gateway-store -p gateway`
- `cargo test -p gateway-service budget_guard::tests -- --nocapture`
- `cargo test -p gateway-store migration -- --nocapture`
- `mise run lint`
- `mise run test`
- `mise run admin-contract-check`
- `mise -C docs run check`

File: crates/admin-ui/web/src/routes/spend-controls.tsx [hunk 1]
```diff
@@ -197,159 +220,270 @@ export function SpendControlsPage() {
...
-              </div>
-            ))}
-          </div>
+                <SelectTrigger>
+                  <SelectValue placeholder="Model" />
+                </SelectTrigger>
+                <SelectContent>
+                  <SelectGroup>
+                    {models.map((model) => (
+                      <SelectItem key={model.id} value={model.id}>
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/admin-ui/web/src/routes/spend-controls.tsx]: Valid. Fixed in 43c339e by exposing the internal model UUID as AdminModelView.model_id and updating Spend Controls to submit that value for model-scoped budget mutations; admin OpenAPI/types were regenerated.

File: crates/admin-ui/web/e2e/gateway-contract.e2e.ts [hunk 2]
```diff
@@ -169,27 +169,28 @@ test('team budget update triggers hard-limit enforcement for service-account key
...
   const budgetsBody = (await budgetsResponse.json()) as {
     data: {
-      teams: Array<{ team_id: string; team_key: string }>
+      service_accounts: Array<{ service_account_id: string; service_account_key: string }>
     }
   }
-  expect(budgetsBody.data.teams.length).toBeGreaterThanOrEqual(1)
-  const teamId = budgetsBody.data.teams[0].team_id
+  expect(budgetsBody.data.service_accounts.length).toBeGreaterThanOrEqual(1)
+  const serviceAccountId = budgetsBody.data.service_accounts[0].service_account_id
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/admin-ui/web/e2e/gateway-contract.e2e.ts:183]: Valid. Fixed in 43c339e: the e2e now selects the seeded service account by service_account_key instead of array position and asserts the upstream stub received zero requests after hard-limit rejection.

File: crates/gateway-core/src/budgets.rs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(tag = "kind", rename_all = "snake_case")]
+pub enum BudgetModelSelector {
+    Model { model_id: Uuid },
+    UpstreamModel { upstream_model: String },
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway-core/src/budgets.rs]: Valid. Fixed in 43c339e by making BudgetModelSelector use trimmed upstream-model identity for equality/serialization/deserialization and reject blank upstream-model selectors on deserialize.

File: crates/gateway-service/src/budget_scopes.rs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                .map(str::trim)
+                .filter(|value| !value.is_empty())
+            {
+                scopes.push(BudgetScope::UserModel {
+                    user_id,
+                    selector: BudgetModelSelector::UpstreamModel {
+                        upstream_model: upstream_model.to_string(),
+                    },
+                });
+            }
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway-service/src/budget_scopes.rs:30]: Verified against issue 106 and the accepted plan. This is intentionally not changed: model_id is authoritative when present, and the trim-only upstream_model fallback applies only when model_id is absent.

File: crates/gateway-store/src/lib.rs [hunk 5]
```diff
@@ -25,22 +25,21 @@ mod tests {
...
+        NewExternalMcpServerRecord, OidcLoginStateRecord, OpenAiCompatDeveloperRole,
+        OpenAiCompatMaxTokensField, OpenAiCompatReasoningEffort, OpenAiCompatRouteCompatibility,
+        PricingCatalogCacheRecord, PricingCatalogRepository, PricingLimits, PricingModalities,
+        PricingProvenance, ProviderCapabilities, RequestAttemptRecord, RequestAttemptStatus,
+        RequestLogPayloadRecord, RequestLogQuery, RequestLogRecord, RequestLogRepository,
+        RequestTag, RequestTags, RequestToolCardinality, RouteCompatibility, SeedApiKey,
+        SeedBudget, SeedModel, SeedModelRoute, SeedProvider, SeedTeam, SeedUser,
+        SeedUserMembership, StoreError, StoreHealth, UpdateExternalMcpServerRecord,
+        UpsertExternalMcpToolRecord, UsageLedgerRecord, UsagePricingStatus, UserStatus,
     };
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway-store/src/lib.rs:43]: Reviewed. The file-size concern is fair as a broader maintainability follow-up, but I am not treating it as a valid blocker for this branch because splitting the existing 6k-line store test module is unrelated to the budget contract correctness fixes and would create a large review surface of its own.

File: crates/gateway-store/src/postgres_store/support.rs [hunk 6]
```diff
@@ -451,77 +451,70 @@ pub(super) fn decode_user_oauth_auth_record(
...
+        },
+        BudgetScopeKind::UserModel => {
+            let user_id = parse_required_uuid(user_id, "user_id")?;
+            let selector = match (model_id, upstream_model) {
+                (Some(model_id), None) => BudgetModelSelector::Model {
+                    model_id: parse_uuid(&model_id)?,
+                },
+                (None, Some(upstream_model)) => BudgetModelSelector::UpstreamModel {
+                    upstream_model: upstream_model.trim().to_string(),
+                },
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway-store/src/postgres_store/support.rs]: Valid. Fixed in 43c339e for both Postgres and libsql decode paths: blank trimmed upstream_model selectors now fail budget record decoding instead of producing an invalid selector.

File: crates/gateway/openapi/admin-api.json [hunk 7]
```diff
@@ -3795,6 +3597,76 @@
...
+            ]
+          },
+          "user_id": {
+            "type": [
+              "string",
+              "null"
+            ]
+          }
+        }
+      },
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway/openapi/admin-api.json:3672]: Valid. Fixed in 43c339e: budget scope request/response schemas are now typed oneOf unions over per-kind shapes, request structs reject unknown fields, and generated admin types/OpenAPI were updated.

File: crates/gateway/src/config.rs [hunk 8]
```diff
@@ -313,9 +313,34 @@ impl GatewayConfig {
...
+                bail!("auth.seed_api_keys entry name cannot be empty");
+            }
+            let service_account = &seed_key.service_account;
+            let service_account_key = normalize_config_service_account_key(&service_account.key)
+                .with_context(|| {
+                    format!("auth.seed_api_keys `{}` service_account.key", seed_key.name)
+                })?;
+            if !seed_service_account_keys.insert(service_account_key.clone()) {
+                bail!("duplicate seeded service account key `{service_account_key}`");
+            }
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway/src/config.rs]: Valid. Fixed in 43c339e: config validation now allows multiple seeded API keys to target the same service account when the normalized service-account name/team/budget settings match, while rejecting conflicting declarations.

File: crates/gateway/src/http/spend.rs [hunk 9]
```diff
@@ -344,282 +313,96 @@ pub async fn list_spend_budgets(
...
-        .store
-        .get_service_account_by_id(service_account_id)
-        .await?
-        .ok_or_else(|| {
-            AppError(GatewayError::InvalidRequest(
-                "service account not found".to_string(),
-            ))
-        })?;
+    let scope = parse_budget_scope(&request.scope)?;
+    validate_budget_scope_exists(&state, &scope).await?;
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway/src/http/spend.rs]: Valid. Fixed in 43c339e: budget deactivation no longer requires the principal row to exist first, so orphaned scopes can be cleaned up. The active service-account key guard remains in place.

File: crates/gateway/src/http/spend.rs [hunk 10]
```diff
@@ -344,282 +313,96 @@ pub async fn list_spend_budgets(
...
+            .count_active_api_keys_for_service_account(service_account_id)
+            .await?
+            > 0
+    {
+        return Err(AppError(GatewayError::InvalidRequest(
+            "cannot deactivate a service account budget while active service account API keys exist"
+                .to_string(),
+        )));
+    }
+    let scope_key = scope.scope_key();
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway/src/http/spend.rs:398]: Verified against the current branch. This is not a bug: the branch compiles, and service_account_id is Copy, so the pattern does not make scope unavailable for the subsequent deactivate call. No code change.

File: crates/gateway/src/http/spend.rs [hunk 11]
```diff
@@ -680,35 +463,149 @@ pub async fn list_budget_alert_history(
...
+                    }
+                }
+                _ => {
+                    return Err(AppError(GatewayError::InvalidRequest(
+                        "user_model budgets require exactly one of model_id or upstream_model"
+                            .to_string(),
+                    )));
+                }
+            };
+            Ok(BudgetScope::UserModel { user_id, selector })
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway/src/http/spend.rs]: Valid. Fixed in 43c339e: user_model budget upsert validation now checks model_id selectors against configured gateway model UUIDs and rejects unknown model IDs.

File: docs/adr/2026-05-27-budget-principal-taxonomy.md [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+Config-seeded API keys declare the service account they create or reconcile, including the owning team and active service-account budget. There is no implicit singleton service account and no reserved `system-legacy` owner.
+
+## Consequences
+
+- Team budget APIs and UI controls are removed.
+- Spend/report filters support `all`, `user`, and `service_account`; `team` is rejected.
+- Team metadata can still narrow or describe service-account reporting, but it is not an ownership scope.
+- Hard-limit enforcement remains best effort under concurrency and may overshoot. Reservations are out of scope.
+
```
[the remark this answers is not in the document]
Comment (ahstn) [reply docs/adr/2026-05-27-budget-principal-taxonomy.md:54]: Valid. Fixed in 43c339e by adding explicit Rationale and Follow-up Items sections to the ADR.

File: docs/adr/2026-05-27-budget-principal-taxonomy.md [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+Budget enforcement is scoped to spend-bearing principals only:
+
+- human users
+- service accounts
+- human user model scopes
+
+Teams are grouping and service-account ownership metadata. Teams are not budget principals, API-key owner kinds, or runtime ownership scopes.
+
+The budget API is generic and scope based:
```
[the remark this answers is not in the document]
Comment (ahstn) [reply docs/adr/2026-05-27-budget-principal-taxonomy.md]: Valid. Fixed in 43c339e by changing the ADR wording to scope-based.

File: crates/gateway-store/src/postgres_store/budgets.rs [hunk 14]
```diff
@@ -1217,28 +919,51 @@ impl BudgetRepository for PostgresStore {
...
+           AND occurred_at < $3"
     );
 
     let row = sqlx::query(query.as_str())
-        .bind(owner_id.to_string())
+        .bind(owner_id)
         .bind(window_start.unix_timestamp())
         .bind(window_end.unix_timestamp())
+        .bind(extra_value)
         .fetch_one(pool)
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway-store/src/postgres_store/budgets.rs]: Valid. Fixed in 2fde27f by only binding the fourth Postgres query parameter when the budget scope predicate actually uses the model/upstream selector placeholder.

File: crates/gateway-service/src/authenticator.rs [hunk 15]
```diff
@@ -82,6 +82,15 @@ where
...
                 return Err(AuthError::ApiKeyOwnerInvalid.into());
             }
+            let scope = BudgetScope::ServiceAccount { service_account_id };
+            if self
+                .repo
+                .get_active_budget_by_scope(&scope)
+                .await?
+                .is_none()
+            {
+                return Err(AuthError::ApiKeyOwnerInvalid.into());
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway-service/src/authenticator.rs]: Valid. Fixed in 2fde27f by adding a dedicated ServiceAccountBudgetRequired auth error that maps to 403/service_account_budget_required, and using it for the service-account active-budget authentication gate.

File: docs/operations/budgets-and-spending.md [hunk 16]
```diff
@@ -29,233 +30,137 @@ Pricing states are explicit:
...
-Use [request-lifecycle-and-failure-modes.md](../reference/request-lifecycle-and-failure-modes.md) for the cross-cutting path from request execution to ledger state.
-
-## Budget Configuration Model
-
-- `user_budgets` stores active and inactive user budgets
-- `team_budgets` stores active and inactive team budgets
-- `service_account_budgets` stores active and inactive service-account budgets
-- each table enforces one active budget per owner
+- post-provider rejection happens before inserting a new priced ledger row
+- duplicate request ids bypass budget math as an idempotent no-op for the same ownership scope
```
[the remark this answers is not in the document]
Comment (ahstn) [reply docs/operations/budgets-and-spending.md:79]: Valid. Fixed in 2fde27f by updating the operator docs to say duplicate request ids are rejected with 400 invalid_request before additional budget math or ledger writes.

File: crates/gateway/src/config.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway/src/config.rs]: Valid, and fixed in 43c339e. Config validation now permits multiple seed API keys for the same service-account key when the normalized service-account declaration matches, while rejecting conflicting declarations.

File: crates/gateway-store/src/libsql_store/budgets.rs [hunk 17]
```diff
@@ -1276,33 +962,54 @@ impl BudgetRepository for LibsqlStore {
...
+           AND occurred_at >= ?2
+           AND occurred_at < ?3"
     );
 
+    let params = libsql::params![
+        owner_id,
+        window_start.unix_timestamp(),
+        window_end.unix_timestamp(),
+        extra_value
+    ];
```
[the remark this answers is not in the document]
Comment (ahstn) [reply crates/gateway-store/src/libsql_store/budgets.rs]: Valid. Fixed in 22baac3 by making the libsql budget spend-sum helper bind three parameters for user/service-account scopes and four only for model/upstream-model scopes.
